### PR TITLE
JP-1845: Update outlier_detection to explicitly handle non-dithered data

### LINKED ISSUE #5613: TSO outlier_detection does not follow requirements
_Issue [JP-1845](https://jira.stsci.edu/browse/JP-1845) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

The algorithm to be used for outlier detection in TSO and Coronagraphy exposures is outlined in <https://outerspace.stsci.edu/display/JWSTCC/Vanilla+Stack+Outlier+Detection.> The current outlier_detection step does **not** follow these specs. Instead, it is using all of the normal imaging mode algorithms, with the simple addition of turning off resampling of the images when forming the median image and hence working in detector space. New functionality that follows the specs needs to be implemented.

### PR DESCRIPTION
**Pull request by karllark**
The outlier_detection step is updated to explicitly do outlier detection differently on non-dithered data versus dithered data.  The non-dithered data only uses the error array to look for outliers.  Dithered data uses the error array and a derivative of the blotted data with the derivative term to handle interpolation issues.

Update the logic on how to pick dithered versus non-dithered data.
Added tests.

General cleanup of the code included.  Removed extraneous calculations in flag_cr not used for JWST.  Added comments and made some variable names more understandable.

More tests with different sized sources needed at a minimum.  And probably updates to the docstrings.

Closes #5613 / [JP-1845](https://jira.stsci.edu/browse/JP-1845)

### CODE REVIEW

**Comment by karllark** [comment]
Ready for review.

**Review by hbushouse — COMMENTED** [review]
So is the "scaled" option and corresponding code no longer relevant or used? IIRC it was only available or applied to TSO imaging data, which might be useful, although there may have been scientific shortcomings in the method, due to its simplicity.

**Comment by karllark** [comment]
> So is the "scaled" option and corresponding code no longer relevant or used? IIRC it was only available or applied to TSO imaging data, which might be useful, although there may have been scientific shortcomings in the method, due to its simplicity.

Yes.  I have not removed the code as it may be useful, but needs to be discussed in the JCalWG before it is re-enabled.  There is another algorithm under discussion that may handle this case better.

File: jwst/outlier_detection/outlier_detection.py [hunk 1]
```diff
@@ -7,10 +7,10 @@
 from scipy import ndimage
 from drizzle.cdrizzle import tblot
 
-from .. import datamodels
-from ..resample import resample
-from ..resample.resample_utils import build_driz_weight, calc_gwcs_pixmap
-from ..stpipe import Step
+from jwst import datamodels
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection.py:10]: I thought that was the best practice so made the change.  I have seen both in the jwst code.  I'm fine with either, just was trying to be a "good" contributor.  Tell me which way to go here and I'll follow.

File: jwst/outlier_detection/outlier_detection.py [hunk 2]
```diff
@@ -301,7 +302,7 @@ def create_median(self, resampled_models):
 
     def blot_median(self, median_model):
         """Blot resampled median image back to the detector images."""
-        interp = self.outlierpars.get('interp', 'poly5')
+        interp = self.outlierpars.get('interp', 'linear')
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection.py:305]: Yes.  Simpler interpolation is usually safer.

File: jwst/outlier_detection/outlier_detection.py [hunk 3]
```diff
@@ -403,96 +400,44 @@ def flag_cr(sci_image, blot_image, **pars):
         # No subtracted background.  Allow user-set value, which defaults to 0
         subtracted_background = backg
 
-    exptime = sci_image.meta.exposure.exposure_time
-
-    sci_data = sci_image.data * exptime
-    blot_data = (blot_image.data + subtracted_background) * exptime
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection.py:409]: It is.  Just needs to be handled differently between the dithered and non-dithered case.

File: jwst/outlier_detection/outlier_detection_step.py [hunk 4]
```diff
@@ -1,17 +1,16 @@
 """Public common step definition for OutlierDetection processing."""
 from functools import partial
 
-from ..stpipe import Step
-from .. import datamodels
+from jwst.stpipe import Step
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection_step.py:4]: Same answer as above.  ;-)

File: jwst/outlier_detection/outlier_detection.py [hunk 5]
```diff
@@ -403,96 +400,44 @@ def flag_cr(sci_image, blot_image, **pars):
...
-    cr_mask = cr_mask.astype(bool)
+    # create the outlier mask
+    if pars.get('resample_data', True):  # dithered outlier detection
+        blot_data += subtracted_background
+        diff_noise = np.abs(sci_data - blot_data)
+
+        # create an initial mask based on
+        # a scaled version of the derivative image (dealing with interpolating issues?)
+        # and the standard n*sigma above the noise
+        t2 = scl1 * blot_deriv + snr1 * err_data
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection.py:420]: The blot_data contribution is to add photon noise to the threshold.  But this is (now) included in the err_data array that includes read, photon, and flat field noise.  Removing it so that the photon noise is not added twice.

File: jwst/outlier_detection/outlier_detection_step.py [hunk 6]
```diff
@@ -54,8 +53,8 @@ class OutlierDetectionStep(Step):
         nhigh = integer(default=0)
         maskpt = float(default=0.7)
         grow = integer(default=1)
-        snr = string(default='4.0 3.0')
-        scale = string(default='0.5 0.4')
+        snr = string(default='5.0 4.0')
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection_step.py:56]: The default `scale` parameters cause failures in the tests for the very unresolved point source test case (single pixel point source).  Previously, this test pasted as the photon noise was included twice in the threshold (I think this is origin at least).  Now that the threshold for detecting an outlier is lower due to not double including the photon noise, the test failed.  I tracked this down to the scale parameters.  I adjusted the defaults for the scale till the tests passed.  Not quite sure why I changed the snr default - probably thought that 5 and 4 sounded better.  I can change the snr defaults back to the original, but will not be able to do so for the scale defaults.

File: jwst/outlier_detection/outlier_detection_step.py [hunk 7]
```diff
@@ -115,30 +114,30 @@ def process(self, input_data):
...
-            elif exptype in TSO_SPEC_MODES:
+            # check for TSO models first
+            if is_tso(single_model):
                 # algorithm selected for TSO data (no resampling)
                 pars['resample_data'] = False  # force resampling off...
                 detection_step = outlier_registry['imaging']
-                pars['resample_suffix'] = 's2d'
-            elif exptype in TSO_IMAGE_MODES + CORON_IMAGE_MODES and \
-                    not self.scale_detection:
+                if exptype in TSO_SPEC_MODES:
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection_step.py:128]: Ok.  Will do.

File: jwst/outlier_detection/outlier_detection_step.py [hunk 8]
```diff
@@ -115,30 +114,30 @@ def process(self, input_data):
...
+                    pars['resample_suffix'] = 'i2d'
+            elif exptype in IMAGE_MODES:
+                # imaging with resampling
+                detection_step = outlier_registry['imaging']
+                pars['resample_suffix'] = 'i2d'
+            elif exptype in CORON_IMAGE_MODES:
                 # algorithm selected for TSO data (no resampling)
                 pars['resample_data'] = False  # force resampling off...
                 detection_step = outlier_registry['imaging']
                 pars['resample_suffix'] = 'i2d'
```
[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection_step.py:140]: Good point.  Done.

File: jwst/outlier_detection/outlier_detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/outlier_detection/outlier_detection.py:10]: @jdavies-st can you comment on the current "wisdom" regarding this issue of absolute vs relative imports? Is there really a preference for one or the other?

[the remark this answers is not in the document]
Comment (jdavies-st) [reply jwst/outlier_detection/outlier_detection.py:10]: PEP8 recommends absolute imports.

https://www.python.org/dev/peps/pep-0008/#imports

But relative imports are acceptable to decrease verbosity.  Karl's changes look fine to me, as none of our absolute import paths are very long relative to the relative import path.

For tests, we should always use absolute imports so we can test against the installed version.

[the remark this answers is not in the document]
Comment (karllark) [reply jwst/outlier_detection/outlier_detection.py:10]: If absolute imports are ok, then this PR is read for re-review/merging from my point-of-view.

[the remark this answers is not in the document]
Comment (jdavies-st) [reply jwst/outlier_detection/outlier_detection.py:10]: Thanks Karl.  Will get you a review today.

File: jwst/outlier_detection/outlier_detection_step.py [hunk 9]
```diff
@@ -115,30 +114,21 @@ def process(self, input_data):
             # Add logic here to select which version of OutlierDetection
             # needs to be used depending on the input data
             if self.input_container:
-                exptype = self.input_models[0].meta.exposure.type
+                single_model = self.input_models[0]
             else:
-                exptype = self.input_models.meta.exposure.type
+                single_model = self.input_models[0]
```
Comment (jdavies-st) [inline jwst/outlier_detection/outlier_detection_step.py:119]: This `if/else` seems to be doing the same thing in both branches.

File: jwst/outlier_detection/tests/test_outlier_detection.py [hunk 10]
```diff
@@ -153,11 +173,91 @@ def test_outlier_step_no_outliers(we_three_sci):
 
 
 def test_outlier_step(we_three_sci):
-    """Test whole step with an outlier"""
+    """Test whole step with an outlier including saving intermediate and results files"""
```
Comment (jdavies-st) [inline jwst/outlier_detection/tests/test_outlier_detection.py:176]: Please use the `_jail` fixture here so that intermediate files are written to  a tmpdir.  I.e.:

```python
def test_outlier_step(_jail, we_three_sci):
    ...
```

File: jwst/outlier_detection/tests/test_outlier_detection.py [hunk 11]
```diff
@@ -153,11 +173,91 @@ def test_outlier_step_no_outliers(we_three_sci):
...
+
+    # Make sure nothing changed in SCI and DQ arrays
+    for image, corrected in zip(container, result):
+        np.testing.assert_allclose(image.data, corrected.data)
+        np.testing.assert_allclose(image.dq, corrected.dq)
+
+
+@pytest.mark.parametrize("exptype", IMAGE_MODES)
+def test_outlier_step_image_weak_CR_dither(exptype):
+    """Test whole step with an outlierf for imaging modes"""
```
Comment (jdavies-st) [inline jwst/outlier_detection/tests/test_outlier_detection.py:222]: sp. "outlier"

**Review by jdavies-st — APPROVED** [review]
Looks good!  Just a few minor comments below.

File: jwst/outlier_detection/outlier_detection_step.py (hunk 9 above)
Comment (karllark) [reply jwst/outlier_detection/outlier_detection_step.py:119]: Good catch.  The else statement should not have `[0]` to match the behavior before this PR.  Fixed.

File: jwst/outlier_detection/tests/test_outlier_detection.py (hunk 10 above)
Comment (karllark) [reply jwst/outlier_detection/tests/test_outlier_detection.py:176]: Will do.  Never used this before, so hopefully just doing as you say is all I need to do.  If there was a pointer to docs on what this does, that would be awesome.

File: jwst/outlier_detection/tests/test_outlier_detection.py (hunk 11 above)
Comment (karllark) [reply jwst/outlier_detection/tests/test_outlier_detection.py:222]: Fixed.

File: jwst/outlier_detection/tests/test_outlier_detection.py (hunk 10 above)
Comment (jdavies-st) [reply jwst/outlier_detection/tests/test_outlier_detection.py]: You think we have time to write documentation for developers?!  😆 

It is documented in the package we get it from

https://ci-watson.readthedocs.io/en/latest/#plugin

but there's no documentation in `jwst` that we actually use this plugin.  😅

**Comment by hbushouse** [comment]
I don't see a change log entry yet (in `CHANGES.rst`)

**Comment by karllark** [comment]
> I don't see a change log entry yet (in `CHANGES.rst`)

Added.

File: jwst/outlier_detection/tests/test_outlier_detection.py (hunk 10 above)
Comment (karllark) [reply jwst/outlier_detection/tests/test_outlier_detection.py]: Neat.  Thanks.

File: CHANGES.rst [hunk 12]
```diff
@@ -113,6 +113,13 @@ master_background
 - Updated documentation to more fully describe the various ways in which the
   step is applied [#5913]
 
+outlier_detection
+-----------------
+
+- outlier detection on non-dithered images is implemented with a simple sigma clipping
+- dithered outlier detection cleaned up and HST specific steps removed
+- additional tests added
+
```
Comment (hbushouse) [inline CHANGES.rst:122]: Please add a reference to the PR number, i.e. [#5822], at the end of each of these lines (or consolidate them into a single entry). See other entries for examples.

Comment (karllark) [reply CHANGES.rst:122]: Done.

**Review by hbushouse — APPROVED** [review]

**Comment by hbushouse** [comment]
Failures in most recent CI run appear to be unrelated.
